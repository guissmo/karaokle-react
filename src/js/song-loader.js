const SUPPORTED_LANGUAGES = ["EN", "TL"];

export const LYRICS_URL = `https://guissmo.com/karaokle-data`;

async function getSongListOfLanguage(lang) {
  const { songArray: songList } = await fetch(
    `${LYRICS_URL}/index_${lang}.json`
  ).then((response) => response.json());
  return songList;
}

export async function whereIsTodaysSong(lang) {
  if (!SUPPORTED_LANGUAGES.includes(lang))
    throw `ERROR: You gave me ${lang} for language.`;
  let timezone = 0;
  if (lang === "FR") timezone = 1;
  if (lang === "EN") timezone = 8;
  if (lang === "TL") timezone = 8;
  const dailyCounter =
    Math.floor((Date.now() + timezone * 3600000) / 86400000) - 19198;
  return whereIsSongWithIndex(lang, dailyCounter);
}

export async function whereIsSongWithIndex(lang, index) {
  const songList = await getSongListOfLanguage(lang);
  return `${LYRICS_URL}/${lang.toLowerCase()}/${songList[index]}.lrc`;
}
