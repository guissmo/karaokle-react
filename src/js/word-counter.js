const FR_PARAMS = {
  preprocess: (x) => {
    return x.replaceAll(/(?<l>s'|m'|t'|c'|j'|l')/gi, `$<l> `);
  },
  wordSpacers: /[ ]/,
  wordReplacers: (word) => word.replaceAll(/[.,?!()]/g, ""),
};

const TL_PARAMS = {
  preprocess: (x) => x,
  wordSpacers: /[ ]/,
  wordReplacers: (word) => word.replaceAll(/[.,?!()]/g, ""),
};

const EN_PARAMS = {
  wordSpacers: " ",
};

const LANG_PARAMS = {
  FR: FR_PARAMS,
  EN: EN_PARAMS,
  TL: TL_PARAMS,
};

export function wordArrFromString(str, lang) {
  const PARAMS = LANG_PARAMS[lang];
  return PARAMS.preprocess(str)
    .split(PARAMS.wordSpacers)
    .map(PARAMS.wordReplacers);
}

export function wordCount(str, lang = "EN") {
  return wordArrFromString(str, lang).length;
}
