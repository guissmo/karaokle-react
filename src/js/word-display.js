export function keepUserAnswerIfCorrect(x) {
  return x.correct ? x.userAnswer : x.correctAnswer;
}

export function getColorArray(x) {
  return x.correct ? "#08ff00" : "#ee0000";
}
