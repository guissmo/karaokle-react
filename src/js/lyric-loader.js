import { LANG_PARAMS } from "./word-counter";

function getFullLyricDataFromLine(line, offset = 0, index) {
  const { min, sec, csec, lyr } = getInfoFromLineViaRegex(line, true);
  return {
    index,
    time: getTimestamp(min, sec, csec, offset),
    lyr: lyr.replaceAll("###", ""),
  };
}

function getStopDataFromLine(line, offset = 0, index) {
  const { min, sec, csec, lyr, stopOffset } = getInfoFromLineViaRegex(line);
  return {
    index,
    time: getTimestamp(min, sec, csec, offset),
    stopTime: getTimestamp(min, sec, csec, offset, stopOffset),
    lyr: lyr.replace(/(\[\[|###).+/g, ""),
  };
}

function handleAlternateSpellings(val, lang) {
  let ret = {};
  const PARAMS = LANG_PARAMS[lang];
  for (let v of val.split(" ")) {
    const words = v.split("|");
    const normalizedCorrectWord = PARAMS.correctionNormalizer(words[0], lang);
    ret[normalizedCorrectWord] = [];
    for (let w of words.slice(1)) {
      ret[normalizedCorrectWord].push(w);
    }
  }
  return ret;
}

function getMetadataFromLine(l, lang) {
  let [key, val] = l.split(":").map((x) => x.trim());
  const ret = {};
  if (key === "lyricOffset" || key === "startAt") val = Number(val);
  if (key === "alternateSpellings") {
    val = handleAlternateSpellings(val, lang);
  }
  ret[key] = val;
  return ret;
}

function getMetadataAndRawLinesFromFile(text) {
  const lines = text.split("\n");
  let passedDashes = false;
  let metadata = {
    lyricOffset: 0,
    alternateSpellings: {},
    startAt: 0,
  }; // default values
  let lyricLines = [];
  for (let l of lines) {
    if (l.trim() === "---") {
      passedDashes = true;
      continue;
    }
    if (!passedDashes) {
      metadata = Object.assign(
        metadata,
        getMetadataFromLine(l, metadata.language)
      );
    } else {
      lyricLines.push(l.trim());
    }
  }
  if (!metadata.videoId) throw "no videoId found";
  if (!metadata.lyricOffset)
    metadata = Object.assign(metadata, {
      lyricOffset: 0,
    });
  return { metadata, lyricLines };
}

export function getInfoFromFile(text) {
  const { metadata, lyricLines } = getMetadataAndRawLinesFromFile(text);

  const noTimestamps = lyricLines
    .map((x) => String(getTimestampViaRegex(x)[1]))
    .reduce((x, y) => `${x.trim()} ${y.trim()}`);

  const fullLyricData = lyricLines.map((x, i) =>
    getFullLyricDataFromLine(x, metadata.lyricOffset, i)
  );

  const roundData = lyricLines
    .map((x, i) => [x, i])
    .filter(([x]) => x.includes("[["))
    .map(([x, i]) => getStopDataFromLine(x, metadata.lyricOffset, i));
  const answers = noTimestamps.matchAll(/\]\][^#]+(?=###)/g);
  let i = 0;
  for (let answer of answers) {
    roundData[i] = {
      ...roundData[i],
      answer: answer[0].replaceAll(/\]\]/g, ""),
    };
    i++;
  }

  return {
    metadata,
    lyricData: {
      full: fullLyricData,
      gapped: roundData,
    },
  };
}

function getTimestamp(min, sec, csec, ...offsets) {
  const totalOffset = offsets.map((x) => Number(x)).reduce((x, y) => x + y, 0);
  return +min * 60 + +sec + +csec / 100 + totalOffset;
}

function getTimestampViaRegex(line) {
  const match = line.match(
    /^\[(?<min>[0-9]+):(?<sec>[0-9]+)\.(?<csec>[0-9]+)\]/
  );
  return [
    {
      min: Number(match.groups.min),
      sec: Number(match.groups.sec),
      csec: Number(match.groups.csec),
    },
    line.replace(match[0], ""),
  ];
}

function getInfoFromLineViaRegex(line, clean = false) {
  const [timeDic, text] = getTimestampViaRegex(line);

  const match = text.match(
    /(?<lyr>.+)\[\[(?<stopOffset>[-.0-9]+)\]\](?<theRest>.*)/
  );

  if (match === null) {
    return {
      ...timeDic,
      lyr: text,
    };
  }

  const { lyr, theRest } = match.groups;
  if (!clean) {
    return {
      ...timeDic,
      ...match.groups,
      lyr: lyr.trim(),
    };
  }

  return {
    ...timeDic,
    lyr: lyr + theRest,
  };
}

/////////////

export async function loadLyricData(songData) {
  const text = await (await fetch(songData)).text();
  return getInfoFromFile(text);
}

export let __testables = {};
if (process.env["NODE_ENV"] === "test")
  __testables = {
    getFullLyricDataFromLine,
    getStopDataFromLine,
    getMetadataAndRawLinesFromFile,
    getInfoFromFile,
    getTimestamp,
    getTimestampViaRegex,
    getInfoFromLineViaRegex,
    handleAlternateSpellings,
  };
