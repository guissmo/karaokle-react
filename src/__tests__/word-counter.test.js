/**
 * @jest-environment jsdom
 */

import { expect, test } from "@jest/globals";
import { wordCount, wordArrFromString } from "../js/word-counter";

test("wordCount - FR", () => {
  const arr = [
    "Je m'présente, je m'appelle Henri",
    "J'ai vu un arc-en-ciel",
    "Puis après je f*rai des galas",
  ];
  expect(wordCount(arr[0], "FR")).toBe(7);
  expect(wordCount(arr[1], "FR")).toBe(5);
  expect(wordCount(arr[2], "FR")).toBe(6);
});

test("wordArrFromString - FR", () => {
  const arr = [
    "Je m'présente, je m'appelle Henri",
    "J'ai vu un arc-en-ciel",
    "Puis après je f'rai des galas",
    "La. La? La! La, la (la)",
    "C'est t'a m'a s'arroser j'ai j'suis l'amour",
  ];
  expect(wordArrFromString(arr[0], "FR")).toStrictEqual([
    "Je",
    "m'",
    "présente",
    "je",
    "m'",
    "appelle",
    "Henri",
  ]);
  expect(wordArrFromString(arr[1], "FR")).toStrictEqual([
    "J'",
    "ai",
    "vu",
    "un",
    "arc-en-ciel",
  ]);
  expect(wordArrFromString(arr[2], "FR")).toStrictEqual([
    "Puis",
    "après",
    "je",
    "f'rai",
    "des",
    "galas",
  ]);
  expect(wordArrFromString(arr[3], "FR")).toStrictEqual([
    "La",
    "La",
    "La",
    "La",
    "la",
    "la",
  ]);
  expect(wordArrFromString(arr[4], "FR")).toStrictEqual([
    "C'",
    "est",
    "t'",
    "a",
    "m'",
    "a",
    "s'",
    "arroser",
    "j'",
    "ai",
    "j'",
    "suis",
    "l'",
    "amour",
  ]);
});

test("wordCount - TL", () => {
  const arr = [
    "Ded na ded talaga ako sa mga pakembot-kembot mo",
    "Basta't 'wag s'yang hawakan 'pag galit",
  ];
  expect(wordCount(arr[0], "TL")).toBe(9);
  expect(wordCount(arr[1], "TL")).toBe(6);
});

test("wordArrFromString - TL", () => {
  const arr = [
    "Ded na ded talaga ako sa mga pakembot-kembot mo",
    "Basta't 'wag s'yang hawakan 'pag galit",
  ];
  expect(wordArrFromString(arr[0], "TL")).toStrictEqual([
    "Ded",
    "na",
    "ded",
    "talaga",
    "ako",
    "sa",
    "mga",
    "pakembot-kembot",
    "mo",
  ]);
  expect(wordArrFromString(arr[1], "TL")).toStrictEqual([
    "Basta't",
    "'wag",
    "s'yang",
    "hawakan",
    "'pag",
    "galit",
  ]);
});
