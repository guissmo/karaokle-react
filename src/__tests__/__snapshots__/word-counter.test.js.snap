// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wordArrFromString - TL 1`] = `
Array [
  Object {
    "correct": true,
    "userAnswer": "DED",
  },
  Object {
    "correct": true,
    "userAnswer": "NA",
  },
  Object {
    "correct": true,
    "userAnswer": "DED",
  },
  Object {
    "correct": false,
    "userAnswer": "tlaaga",
  },
  Object {
    "correct": true,
    "userAnswer": "ako",
  },
  Object {
    "correct": true,
    "userAnswer": "sa",
  },
  Object {
    "correct": true,
    "userAnswer": "MGA",
  },
  Object {
    "correct": true,
    "userAnswer": "pakembotkembot",
  },
  Object {
    "correct": true,
    "userAnswer": "mo",
  },
]
`;

exports[`wordArrFromString - TL 2`] = `
Array [
  Object {
    "correct": true,
    "userAnswer": "Basta't",
  },
  Object {
    "correct": true,
    "userAnswer": "wag",
  },
  Object {
    "correct": true,
    "userAnswer": "syang",
  },
  Object {
    "correct": true,
    "userAnswer": "hawakan",
  },
  Object {
    "correct": true,
    "userAnswer": "pag",
  },
  Object {
    "correct": true,
    "userAnswer": "galit",
  },
]
`;

exports[`wordArrFromString - TL 3`] = `
Array [
  Object {
    "correct": true,
    "userAnswer": "nasa",
  },
  Object {
    "correct": true,
    "userAnswer": "isip",
  },
  Object {
    "correct": true,
    "userAnswer": "ko",
  },
  Object {
    "correct": false,
    "userAnswer": "siya",
  },
  Object {
    "correct": false,
    "userAnswer": "araw-araw",
  },
  Object {
    "correct": false,
    "userAnswer": "aking",
  },
  Object {
    "correct": false,
    "userAnswer": "sinta",
  },
  Object {
    "correct": false,
    "userAnswer": null,
  },
]
`;

exports[`wordComparison - FR 1`] = `
Array [
  Object {
    "correct": true,
    "userAnswer": "Je",
  },
  Object {
    "correct": true,
    "userAnswer": "me",
  },
  Object {
    "correct": true,
    "userAnswer": "presente",
  },
  Object {
    "correct": true,
    "userAnswer": "je",
  },
  Object {
    "correct": true,
    "userAnswer": "m'",
  },
  Object {
    "correct": true,
    "userAnswer": "appelle",
  },
  Object {
    "correct": true,
    "userAnswer": "Henri",
  },
]
`;

exports[`wordComparison - FR 2`] = `
Array [
  Object {
    "correct": true,
    "userAnswer": "j'",
  },
  Object {
    "correct": true,
    "userAnswer": "ai",
  },
  Object {
    "correct": true,
    "userAnswer": "vu",
  },
  Object {
    "correct": true,
    "userAnswer": "un",
  },
  Object {
    "correct": true,
    "userAnswer": "arc-en-ciel",
  },
]
`;

exports[`wordComparison - FR 3`] = `
Array [
  Object {
    "correct": true,
    "userAnswer": "PUIS",
  },
  Object {
    "correct": true,
    "userAnswer": "APRES",
  },
  Object {
    "correct": true,
    "userAnswer": "JE",
  },
  Object {
    "correct": true,
    "userAnswer": "FERAI",
  },
  Object {
    "correct": true,
    "userAnswer": "DES",
  },
  Object {
    "correct": true,
    "userAnswer": "GALAS",
  },
]
`;

exports[`wordComparison - FR 4`] = `
Array [
  Object {
    "correct": true,
    "userAnswer": "LA",
  },
  Object {
    "correct": true,
    "userAnswer": "LA",
  },
  Object {
    "correct": true,
    "userAnswer": "LA",
  },
  Object {
    "correct": true,
    "userAnswer": "LA",
  },
  Object {
    "correct": true,
    "userAnswer": "LA",
  },
  Object {
    "correct": true,
    "userAnswer": "LA",
  },
]
`;

exports[`wordComparison - FR 5`] = `
Array [
  Object {
    "correct": true,
    "userAnswer": "c'",
  },
  Object {
    "correct": true,
    "userAnswer": "Est",
  },
  Object {
    "correct": true,
    "userAnswer": "t'",
  },
  Object {
    "correct": true,
    "userAnswer": "A",
  },
  Object {
    "correct": true,
    "userAnswer": "m'",
  },
  Object {
    "correct": true,
    "userAnswer": "A",
  },
  Object {
    "correct": true,
    "userAnswer": "s'",
  },
  Object {
    "correct": true,
    "userAnswer": "Arroser",
  },
  Object {
    "correct": true,
    "userAnswer": "j'",
  },
  Object {
    "correct": true,
    "userAnswer": "Ai",
  },
  Object {
    "correct": true,
    "userAnswer": "j'",
  },
  Object {
    "correct": true,
    "userAnswer": "suis",
  },
  Object {
    "correct": true,
    "userAnswer": "l'",
  },
  Object {
    "correct": true,
    "userAnswer": "AMOUR",
  },
]
`;
